HLTrigger/btau/src/HLTDisplacedtktktkVtxProducer.cc

```diff
@@ -0,0 +1,280 @@
+	
+  //get the transient track builder:
+  edm::ESHandle<TransientTrackBuilder> theB;
+  iSetup.get<TransientTrackRecord>().get("TransientTrackBuilder",theB);
+
+  std::unique_ptr<VertexCollection> vertexCollection(new VertexCollection());
+
+  // look at all trackcands,  check cuts and make vertices
+  double e1,e2,e3;
+  Particle::LorentzVector p,p1,p2,p3,pres;
+
+  if ( trackcands->size()  < 3 )   return;
+	
+  RecoChargedCandidateCollection::const_iterator cand1;
+  RecoChargedCandidateCollection::const_iterator cand2;
+  RecoChargedCandidateCollection::const_iterator cand3;
+	
+  // get the objects passing the previous filter
+  Handle<TriggerFilterObjectWithRefs> previousCands;
+  iEvent.getByToken(previousCandToken_,previousCands);
+
+  vector<RecoChargedCandidateRef> vPrevCands;
+  previousCands->getObjects(triggerTypeDaughters_,vPrevCands);
+
+  std::vector<bool> candComp;
+  for (cand1=trackcands->begin(); cand1!=trackcands->end(); cand1++)
+    candComp.push_back(checkPreviousCand( cand1->get<TrackRef>(), vPrevCands));	
+  
+  for (cand1=trackcands->begin(); cand1!=trackcands->end(); cand1++) {
+
+    TrackRef tk1 = cand1->get<TrackRef>();
+    LogDebug("HLTDisplacedtktktkVtxProducer") << " 1st track in loop: q*pt= " << cand1->charge()*cand1->pt() << ", eta= " << cand1->eta() << ", hits= " << tk1->numberOfValidHits();
+	     
+    //first check if this track passed the previous filter
+    if (!candComp[cand1-trackcands->begin()]) continue;
+    // if( ! checkPreviousCand( tk1, vPrevCands) ) continue;
+    
+    // cuts
+    if (abs(cand1->eta())>maxEta_) continue;
+    if (cand1->pt() < firstTrackPt) continue;
+    
+    cand2=trackcands->begin();
+    if(firstTrackMass==secondTrackMass){cand2 = cand1+1;}
+
+    for (; cand2!=trackcands->end(); cand2++) {
+		     
+      TrackRef tk2 = cand2->get<TrackRef>();
+      if(tk1==tk2) continue;		 
+      LogDebug("HLTDisplacedtktktkVtxProducer") << " 2nd track in loop: q*pt= " << cand2->charge()*cand2->pt() << ", eta= " << cand2->eta() << ", hits= " << tk2->numberOfValidHits() << ", d0= " << tk2->d0();
+
+      //first check if this track passed the previous filter
+      if (!candComp[cand2-trackcands->begin()]) continue;
+      // if( ! checkPreviousCand( tk2, vPrevCands) ) continue;
+			 
+      // cuts
+      if (abs(cand2->eta()) > maxEta_) continue;
+      if (cand2->pt() < secondTrackPt) continue;
+
+      // opposite sign or same sign for resonance
+      if(resOpt_>0)
+        {
+          if (chargeOpt_<0) {
+            if (cand1->charge()*cand2->charge()>0) continue;
+          } else if (chargeOpt_>0) {
+            if (cand1->charge()*cand2->charge()<0) continue;
+          }
+        }
+
+      // 	    	    
+      // Combined ditrack system
+      e1 = sqrt(cand1->momentum().Mag2()+firstTrackMass2);
+      e2 = sqrt(cand2->momentum().Mag2()+secondTrackMass2);
+      p1 = Particle::LorentzVector(cand1->px(),cand1->py(),cand1->pz(),e1);
```

Comment:
the p1 and p2 temporary creation can be avoided.

Reply:
changed